ls1intum/Artemis
# [Bugfix] Preliminary result is not always correct

### Checklist

### Motivation and Context


According to #2761, the result stays falsely as preliminary if the Build After Due Date is set.  


### Description

The result is checked against its completion date and the assessment due date or the Build After Due Date.
If the programming exercise has a semi-automatic assessment type, the result stays preliminary as long as either the due date is not over or (in case no due date is set) the assessment type of the latest result is still automatic.

fixes #2761 

### Steps for Testing


1. Log in to Artemis
2. Create a programming exercise with a due date and ..
2.1 a build after due date, activate manual review and an assessment due date
2.2 a build after due date, activate manual review and NO assessment due date
2.3 a build after due date and NO manual review
2.4 NO build after due date, activate manual review and and assessment due date
2.5 NO build after due date, activate manual review and NO assessment due date
2.6 NO build after due date and NO manual review
3. Log in as a student and participate in these exercises
4. The result string should either have a `(preliminary)` attached at the end + tooltip or not
4.1 Verify that it is not attached for 2.6 but for 2.1 - 2.5
4.2 Verify that the tooltip varies whether manual review is activated or not
4.3 Verify that after the last activated due date is over, that the `(preliminary)` is gone

### Test Coverage



No relevant changes to the functionality of the code.

### Screenshots



Settings for variant 2.1:
![image](https://user-images.githubusercontent.com/33484318/106123675-89190000-615a-11eb-9432-4e181adc4d7f.png)

Course Overview:
![image](https://user-images.githubusercontent.com/33484318/106123438-35a6b200-615a-11eb-9fcc-85d1bc4ca3e2.png)

Online Editor (build result):
![image](https://user-images.githubusercontent.com/33484318/106123395-2889c300-615a-11eb-8fb6-af5a7ff8c7d2.png)


Showing preliminary only for 2.1-2.5:
![image](https://user-images.githubusercontent.com/33484318/106125980-19f0db00-615d-11eb-95dc-69479f42fd30.png)


Tooltip for 2.3:
![image](https://user-images.githubusercontent.com/33484318/106125692-bb2b6180-615c-11eb-9ec2-24999c7bcc0f.png)

Tooltip for the rest: 
![image](https://user-images.githubusercontent.com/33484318/106125793-dac28a00-615c-11eb-8ed3-aa63131a32cb.png)

Build after due date is over (2.3 is not preliminary anymore):
![image](https://user-images.githubusercontent.com/33484318/106127740-16f6ea00-615f-11eb-8fdd-bc4fb57d3214.png)

Assessment due date is over (those without an assessment due date are still preliminary):
![image](https://user-images.githubusercontent.com/33484318/106129084-3c382800-6160-11eb-8c86-53d0398c4ad9.png)

After manual assessment:
![image](https://user-images.githubusercontent.com/33484318/106129538-a0f38280-6160-11eb-88f6-35626a17dddc.png)
